File: src/skillguard/core/rules_loader.py
"""Load detection rules from YAML files."""

from __future__ import annotations

from pathlib import Path

import yaml

from skillguard.core.models import DetectionRule

# Default rules directory (relative to package root)
_DEFAULT_RULES_DIR = Path(__file__).resolve().parent.parent.parent.parent / "rules"


def load_rules(
    rules_dir: str | Path | None = None,
    engine_filter: str | None = None,
    category_filter: str | None = None,
    enabled_only: bool = True,
) -> list[DetectionRule]:
    """Load detection rules from YAML files in the rules directory.

    Args:
        rules_dir: Path to rules directory. Uses default if None.
        engine_filter: Only load rules for this engine (e.g., 'REGEX').
        category_filter: Only load rules for this category.
        enabled_only: Only load enabled rules.

    Returns:
        List of DetectionRule objects.
    """
    path = Path(rules_dir) if rules_dir else _DEFAULT_RULES_DIR
    if not path.exists():
        return []

    rules: list[DetectionRule] = []
    for yml_file in sorted(path.rglob("*.yml")):
        try:
            rule = _load_rule_file(yml_file)
        except Exception:
            continue

        if rule is None:
            continue
        if enabled_only and not rule.enabled:
            continue
        if engine_filter and rule.engine.upper() != engine_filter.upper():
            continue
        if category_filter and rule.category != category_filter:
            continue

        rules.append(rule)

    return rules


def _load_rule_file(path: Path) -> DetectionRule | None:
    """Load a single rule YAML file."""
    raw = path.read_text(encoding="utf-8")
    data = yaml.safe_load(raw)
    if not isinstance(data, dict) or "id" not in data:
        return None

    return DetectionRule(
        id=data["id"],
        name=data.get("name", data["id"]),
        description=data.get("description", ""),
        severity=data.get("severity", "medium"),
        category=data.get("category", "unknown"),
        owasp_llm=data.get("owasp_llm", []),
        mitre_attack=data.get("mitre_attack", []),
        target=data.get("target", "ANY"),
        engine=data.get("engine", "REGEX"),
        pattern=data.get("pattern", ""),
        false_positive_notes=data.get("false_positive_notes"),
        remediation=data.get("remediation"),
        references=data.get("references", []),
        enabled=data.get("enabled", True),
    )

File: go-skillguard/internal/core/rules_loader.go
package core

import (
	"os"
	"path/filepath"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"
)

// LoadRules loads detection rules from YAML files in the rules directory.
func LoadRules(rulesDir string, engineFilter string, categoryFilter string, enabledOnly bool) ([]DetectionRule, error) {
	if rulesDir == "" {
		return []DetectionRule{}, nil
	}

	info, err := os.Stat(rulesDir)
	if err != nil || !info.IsDir() {
		return []DetectionRule{}, nil
	}

	var paths []string
	err = filepath.Walk(rulesDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if !info.IsDir() && strings.HasSuffix(strings.ToLower(path), ".yml") {
			paths = append(paths, path)
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	sort.Strings(paths)

	var rules []DetectionRule
	for _, p := range paths {
		rule, err := loadRuleFile(p)
		if err != nil || rule == nil {
			continue
		}
		if enabledOnly && !rule.Enabled {
			continue
		}
		if engineFilter != "" && !strings.EqualFold(rule.Engine, engineFilter) {
			continue
		}
		if categoryFilter != "" && rule.Category != categoryFilter {
			continue
		}
		rules = append(rules, *rule)
	}

	return rules, nil
}

func loadRuleFile(path string) (*DetectionRule, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var raw map[string]any
	if err := yaml.Unmarshal(data, &raw); err != nil {
		return nil, err
	}

	if raw == nil || raw["id"] == nil {
		return nil, nil
	}

	rule := &DetectionRule{
		ID:          getString(raw, "id"),
		Name:        getStringDefault(raw, "name", getString(raw, "id")),
		Description: getStringDefault(raw, "description", ""),
		Severity:    Severity(getStringDefault(raw, "severity", "medium")),
		Category:    getStringDefault(raw, "category", "unknown"),
		OWASPLLM:    getStringSlice(raw, "owasp_llm"),
		MITREAttack: getStringSlice(raw, "mitre_attack"),
		Target:      getStringDefault(raw, "target", "ANY"),
		Engine:      getStringDefault(raw, "engine", "REGEX"),
		Pattern:     raw["pattern"],
		FalsePositiveNotes: getStringDefault(raw, "false_positive_notes", ""),
		Remediation: getStringDefault(raw, "remediation", ""),
		References:  getStringSlice(raw, "references"),
		Enabled:     getBoolDefault(raw, "enabled", true),
	}

	return rule, nil
}

func getString(m map[string]any, key string) string {
	if v, ok := m[key]; ok {
		if s, ok := v.(string); ok {
			return s
		}
	}
	return ""
}

func getStringDefault(m map[string]any, key, def string) string {
	s := getString(m, key)
	if s == "" {
		return def
	}
	return s
}

func getStringSlice(m map[string]any, key string) []string {
	v, ok := m[key]
	if !ok || v == nil {
		return nil
	}
	if arr, ok := v.([]any); ok {
		result := make([]string, 0, len(arr))
		for _, item := range arr {
			if s, ok := item.(string); ok {
				result = append(result, s)
			}
		}
		return result
	}
	return nil
}

func getBoolDefault(m map[string]any, key string, def bool) bool {
	v, ok := m[key]
	if !ok {
		return def
	}
	if b, ok := v.(bool); ok {
		return b
	}
	return def
}
